qiskit/user_config.py

```diff
@@ -138,6 +139,67 @@ def read_config_file(self):
                 self.settings['num_processes'] = num_processes
 
 
+def set_config(key, value, section=None, file=None):
+    """Adds or modifies a user configuration
+
+    It will add the configuration in either the default location
+    ~/.qiskit/settings.conf or the value of file argument.
+
+    Only valid user config can be set in 'default' section. Custom
+    user config can be added in any other sections.
+
+    Args:
+        key (str): name of the config
+        value (str): value of the config
+        section (str, optional): if not specified, adds it to the
+            `default` section of the config file.
+        file (str, optional): the file to which config is added.
+            If not specified, adds it to the default config file.
+
+    Raises:
+        QiskitUserConfigError: if the config is invalid
+    """
+    filename = file or DEFAULT_FILENAME
+    sectname = 'default' if section is None else section
+
+    if not isinstance(key, str):
+        raise exceptions.QiskitUserConfigError('Key must be string type')
+    if not isinstance(value, str):
+        raise exceptions.QiskitUserConfigError('Value must be string type')
+
+    valid_config = ['circuit_drawer',
+                    'circuit_mpl_style',
+                    'circuit_mpl_style_path',
+                    'transpile_optimization_level',
+                    'parallel',
+                    'num_processes']
+
+    if section in [None, 'default']:
+        if key not in valid_config:
+            raise exceptions.QiskitUserConfigError(
+                "{} is not a valid user config.".format(key))
+
+    config = configparser.ConfigParser()
+    config.read(filename)
+
+    if sectname not in config.sections():
+        config.add_section(sectname)
+
+    config.set(sectname, key, value)
+
+    try:
+        with open(filename, 'w') as cfgfile:
+            config.write(cfgfile)
+    except OSError as ex:
+        raise exceptions.QiskitUserConfigError(
+            "Unable to load the config file {}. Error: '{}'".format(
+                filename, str(ex)))
+
+    filename = os.getenv('QISKIT_SETTINGS', DEFAULT_FILENAME)
+    user_config = UserConfig(filename)
+    user_config.read_config_file()
```

Comment:
Is this for validation? We should have a comment here about the purpose of this. Although if it is for validation we shouldn't we be using the filename input if it's present?

Reply:
The previous line sets `user_config.filename`, which is the file that `read_config_file` reads.

Reply:
this was actually the whole block, not just line 200. The issue is that `filename` is set in L198 to `os.getenv('QISKIT_SETTINGS', DEFAULT_FILENAME)` which means if the users passes in a filename the validation won't be run on it, instead it will be validating the loaded file for the session not the new one.

Reply:
But it will be temporary if we set the config from the file user passes. 
`__init__.py` calls a `get_config` function, which sets the config file to `QISKIT_SETTINGS` env or default config file.
```
def get_config():
    filename = os.getenv('QISKIT_SETTINGS', DEFAULT_FILENAME)
    user_config = UserConfig(filename)
    user_config.read_config_file()
    return user_config.settings
```
So next time there is an `import qiskit` statement, the config file is reset. This config change might confuse the user right?

Reply:
This is just for validation right? There is no expectation that this is changing the loaded config file. Right now we do no validation if the user specifies an different output path then the configured file. If we're going to validate the output we should do it for all cases. For example right now if you do something like:

```python
from qiskit.user_config import set_config

# Generate a new config file
set_config('transpile_optimization_level', '42', file_path='new_file.ini')
```
it will work fine which is not desireable because when I go to use `new_file.ini` later it will fail to load.

However if you do:


```python
from qiskit.user_config import set_config

# Generate a new config file
set_config('transpile_optimization_level', '42')
```

it will correctly fail. We need to be validating the generated file in all cases.